### PR TITLE
GODRIVER-2191 Allow connection handshake errors during checkOut to be retried.

### PR DESCRIPTION
**Pull request by matthewdale**
[GODRIVER-2191](https://jira.mongodb.org/browse/GODRIVER-2191)

Return a retryable `ConnectionError` from `checkOut()` if the connection handshake fails. Sync all related spec tests.

Changes:
* Allow `ConnectionError` to satisfy the `driver.RetryablePoolError` interface and return `retryable = true` for connection initialization and handshake errors.
* Sync the "retryable-reads", "retryable-writes", and "transactions" spec tests.
* Update the "retryable-reads" test paths in the unified and legacy spec test runners.
* Support the CMAP event type "connectionCheckOutStartedEvent" in the unified spec test runner.

### CODE REVIEW

**Review by benjirewis — APPROVED** [review]

File: x/mongo/driver/topology/errors.go [hunk 1]
```diff
@@ -4,8 +4,12 @@ import (
 	"fmt"
 
 	"go.mongodb.org/mongo-driver/mongo/description"
+	"go.mongodb.org/mongo-driver/x/mongo/driver"
 )
 
+// Assert that ConnectionError satisfies driver.RetryablePoolError.
+var _ driver.RetryablePoolError = ConnectionError{}
```
Comment (kevinAlbs) [inline x/mongo/driver/topology/errors.go:11]: The Retryable Writes spec says:

> When the driver encounters a network error establishing an initial connection to a server,
  it MUST add a RetryableWriteError label to that error if the MongoClient performing
  the operation has the retryWrites configuration option set to true.

IIUC, the `RetryableWriteError` is already applied to network errors that occur during the handshake commands.
The [handshaker](https://github.com/mongodb/mongo-go-driver/blob/9101b419216336cec091fc70d1f622efa12f5ca0/mongo/client.go#L435-L439) will eventually call `operation.Execute`.

Does this need to be applied to other network errors? For example, if [configureTLS](https://github.com/mongodb/mongo-go-driver/blob/9101b419216336cec091fc70d1f622efa12f5ca0/x/mongo/driver/topology/connection.go#L197-L200) returns an error?

**Review by kevinAlbs — COMMENTED** [review]
Left an open question about applying the `RetryableWriteError` label and a concern that not all `ConnectionError` should be considered retryable.

**Review by matthewdale — COMMENTED** [review]
Based on your feedback, I'm not sure if it's possible to satisfy the Retryable Reads and Retryable Writes specs with the current retry mechanism and the information available while establishing a new connection. I will investigate what's possible, but may have to refactor the operation retry logic.

**Comment by matthewdale** [comment]
After considering @kevinAlbs's feedback, I've moved this back to a draft PR and am reconsidering how to implement this.

**Comment by matthewdale** [comment]
Closing this for now since we're deferring this task.

**Pull request closed without merging**